src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -0,0 +1,164 @@
+        except StopIteration:
+            return False
+        else:
+            return True
+
+    @classmethod
+    @abstractmethod
+    def get_local_implementation(cls) -> "RpcService":
+        """Return a service object that runs locally.
+
+        The returned service object is (generally) the database-backed instance that
+        is called when we either receive a remote call from outside, or want to call
+        it within the same silo.
+
+        A base service class generally should override this class method, making a
+        forward reference to its own database-backed subclass.
+        """
+
+        raise NotImplementedError
+
+    @classmethod
+    def _create_local_delegation(cls) -> "RpcService":
+        """Create a lazy wrapper around the local implementation.
+
+        The indirection is necessary so that a base service's
+        `get_local_implementation` body can safely make a forward reference that
+        `resolve_to_delegation` won't call yet.
+        """
+        impl_attrname = "__cached_local_implementation"
+
+        @cached_property  # type: ignore
+        def cached_local_implementation(service_obj: "RpcService") -> "RpcService":
+            return service_obj.get_local_implementation()
+
+        def create_delegating_method(base_method: Callable[..., Any]) -> Callable[..., Any]:
+            method_name = base_method.__name__
+
+            def delegating_method(service_obj: "RpcService", **kwargs: Any) -> Any:
+                impl = getattr(service_obj, impl_attrname)
+                method = getattr(impl, method_name)
+                return method(**kwargs)
+
+            return delegating_method
+
+        overrides = {
+            service_method.__name__: create_delegating_method(service_method)
+            for service_method in cls._get_all_abstract_rpc_methods()
+        }
+        overrides[impl_attrname] = cached_local_implementation
+        remote_service_class = type(f"{cls.__name__}__LocalDelegate", (cls,), overrides)
+        return cast(RpcService, remote_service_class())
+
+    @classmethod
+    def _create_remote_delegation(cls) -> "RpcService":
+        """Create a service object that makes remote calls to another silo."""
+
+        def create_remote_method(base_method: Callable[..., Any]) -> Callable[..., Any]:
+            method_name = base_method.__name__
+
+            def remote_method(**kwargs: Any) -> Any:
+                # TODO: Handle kwarg serialization
+                return dispatch_remote_call(cls.name, method_name, kwargs)
+
+            return remote_method
+
+        overrides = {
+            service_method.__name__: create_remote_method(service_method)
+            for service_method in cls._get_all_abstract_rpc_methods()
+        }
+        remote_service_class = type(f"{cls.__name__}__RemoteDelegate", (cls,), overrides)
+        return cast(RpcService, remote_service_class())
+
+    @classmethod
+    def resolve_to_delegation(cls) -> "RpcService":
```

Comment:
In the end, some services won't be RPC backend -- instead, there will be two competing local implementations.

For instance, the intention is to split the `organization_service` into two services: `organization_service` which backs the RPC -> full organization lookup, and the `organization_mapping_service` which actually supports local lookups of the summarized mapping data for both region and control silo without RPC (for instance, knowing if a user belongs to an org won't require an RPC because a different model on each silo will support that locally).  We don't have many examples of this yet, but as we work on performance, we likely will end up replicating models and supporting services that alternate their local implementations.  It's a special case, but one I think we should be mindful of.

For now, perhaps we should simply name this method to make it clear that it is specifically for remote <-> local delegation, as opposed to local, local delegation.

Reply:
Admittingly this *is* scoped to `RpcService`, but I do still ponder if adding something else to the method name is super clear and helpful too?  Unless we intend for method overloading semantic to help us, but I ponder.

Reply:
Actually, disregard.  This is fine.  We'll address what the non Rpc Service approach looks like when we get there instead.

Reply:
Yeah, I think we're on the same page. This is, for better or worse, introducing a new section to the Venn diagram: not all hybrid cloud services are RPC services, but most are. There's room for hybrid cloud service classes that *don't* extend `RpcService`, and I intentionally left alone code such as

https://github.com/getsentry/sentry/blob/cd2e8f6e50aecb291f11530d3dcfc997fada24c0/src/sentry/services/hybrid_cloud/log/__init__.py#L59-L65

and

https://github.com/getsentry/sentry/blob/cd2e8f6e50aecb291f11530d3dcfc997fada24c0/src/sentry/services/hybrid_cloud/tombstone/__init__.py#L42-L48

However, I wouldn't mind unifying some of the new code in `rpc.py` with the preexisting scaffolding.